### PR TITLE
[gw-api] publish supported features in gateway class

### PR DESCRIPTION
**Pull request by LiorLieberman**
**Please provide a description of this PR:**
This PR implements [GEP-2162](https://github.com/kubernetes-sigs/gateway-api/blob/f80f447fbf6e759a196cab6532b70582d9a8bd70/geps/gep-2162/index.md). (I hopefully added it on the right place)

I am interested in opinions about how to incorporate the tests for it. Adding a static list for a the golden.yaml files seems sub-optimal as we will need to change this every time we support a new feature. 
Also it feels like a lot of duplication.

feedback and ideas are welcomed. 

- [X] Gateway API

### CODE REVIEW

File: pilot/pkg/config/kube/gateway/model.go
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/model.go:42]: nit: you can make this a static var instead of computing it each time `var supported = func() []k8sv1.SupportedFeature { ... }()`

File: pilot/pkg/config/kube/gateway/model.go
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/model.go:40]: For tests...

> I am interested in opinions about how to incorporate the tests for it. Adding a static list for a the golden.yaml files seems sub-optimal as we will need to change this every time we support a new feature.

Maybe we can override to 1-2 features in the test so we don't overload it. We don't really need to test that we write _all_ of them, just that we write any (presumably, if we write some we could write them all).

---

I am a bit worried about our code hardcoding AllFeatures. Should we have this somewhere that this code and our conformance test (tests/integration/pilot/gateway_conformance_test.go) read from? That way they cannot get out of sync

File: pilot/pkg/config/kube/gateway/conversion.go
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/gateway/conversion.go:1855]: Not sure why we need it.

File: pilot/pkg/config/kube/gateway/conversion.go
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/gateway/conversion.go:1969]: And i think we donot have to calculate it each time, we can  declare a variable.

File: pilot/pkg/config/kube/gateway/model.go
Comment (LiorLieberman) [reply pilot/pkg/config/kube/gateway/model.go:40]: Ultimately, and perhaps relatively soon after this is merged(or together with this change if you think its better), we want to take the list of supportedFeatures from GatewayClass, and based on this list run the conformance tests.

Regardless, I agree that this should live in somewhere appropriate in the codebase. I thought model.go is the file to go but perhaps I could add a new file. 
wdyt?

File: pilot/pkg/config/kube/gateway/conversion.go
Comment (LiorLieberman) [reply pilot/pkg/config/kube/gateway/conversion.go:1855]: https://github.com/kubernetes-sigs/gateway-api/issues/2162

File: pilot/pkg/config/kube/gateway/conversion.go
Comment (LiorLieberman) [reply pilot/pkg/config/kube/gateway/conversion.go:1969]: thanks, changed

File: pilot/pkg/config/kube/gateway/model.go
Comment (LiorLieberman) [reply pilot/pkg/config/kube/gateway/model.go:42]: changed thanks

**Review by howardjohn — APPROVED** [review]

**Comment by LiorLieberman** [comment]
Any ideas why the tests pods continuously fail to initialize ? it blocks this PR

File: pilot/pkg/config/kube/gateway/supported_features.go
Comment (levikobi) [inline pilot/pkg/config/kube/gateway/supported_features.go:24]: I think this might be your problem. `suite.AllFeatures` contains elements that aren't part of the kubebuilder validation list (see [here](https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1/gatewayclass_types.go#L269)).
I'm currently facing this problem as I try to implement `SupportedFeatures` at envoy.

I haven't found a list of defined supported features in gateway-api's that can filter out things from `suite.AllFeatures` that aren't in the validation list. I think it can be helpful to add such a list.

Comment (LiorLieberman) [reply pilot/pkg/config/kube/gateway/supported_features.go:24]: So far it seems to be it, thanks @levikobi !

**Comment by LiorLieberman** [comment]
Ok, I am leaving the last commit temporarily until we change kubebuilder validation on GWC status. 
/hold

/cc @robscott

**Comment by LiorLieberman** [comment]
https://github.com/kubernetes-sigs/gateway-api/pull/2754 to fix the issue upstream.

**Review by howardjohn — APPROVED** [review]

**Pull request merged**
